### PR TITLE
Add index signature to IResolvers and IDirectiveResolvers

### PR DESCRIPTION
**Pull request by ardatan**
Related to  #1133

### CODE REVIEW

File: packages/plugins/typescript-resolvers/src/root.handlebars [hunk 1]
```diff
@@ -79,10 +79,10 @@ export interface IResolvers<TContext = {{{ getContext }}}> {
   {{#each scalars}}
     {{ convert name 'typeNames'}}{{#unless @root.config.strict}}?{{/unless}}: GraphQLScalarType;
   {{/each}}
-}
+} & { [typeName: string] : never };
```
Comment (kamilkisiela) [inline packages/plugins/typescript-resolvers/src/root.handlebars]: Are we sure it won't mark all properties as `never`?

File: packages/plugins/typescript-resolvers/tests/typescript-resolvers.spec.ts [hunk 2]
```diff
@@ -1230,23 +1230,23 @@ describe('Resolvers', () => {
...
         Node?: NodeResolvers;
         PostOrUser?: PostOrUserResolvers;
         Date?: GraphQLScalarType;
-      }
+      } & { [typeName: string] : never };
     `);
 
     expect(content).toBeSimilarStringTo(`
-      export interface IDirectiveResolvers<Result> {
+      export interface IDirectiveResolvers<Result> = {
```
Comment (kamilkisiela) [inline packages/plugins/typescript-resolvers/tests/typescript-resolvers.spec.ts]:
```suggestion
      export type IDirectiveResolvers<Result> = {
```

File: packages/plugins/typescript-resolvers/src/root.handlebars (hunk 1 above)
Comment (dotansimha) [reply packages/plugins/typescript-resolvers/src/root.handlebars]: Yeah looks suspicious

File: packages/plugins/typescript-resolvers/tests/typescript-resolvers.spec.ts (hunk 2 above)
Comment (dotansimha) [reply packages/plugins/typescript-resolvers/tests/typescript-resolvers.spec.ts]: Yeah `type` seems better :)

**Comment by dotansimha** [comment]
Please note @kamilkisiela 's notes, and fix CI.

File: packages/plugins/typescript-resolvers/src/root.handlebars (hunk 1 above)
Comment (ardatan) [reply packages/plugins/typescript-resolvers/src/root.handlebars]: You can try it :)

**Comment by ardatan** [comment]
@dotansimha Fixed 👍

**Comment by SimenB** [comment]
How are you supposed to implement this interface?

See minimal example: https://github.com/SimenB/ts-codegen-resolver
It has a type error when trying to create an `IResolver`

What am I doing wrong?

**Comment by domasx2** [comment]
TS 3.3.333 this interfaces are now unimplementable, 'not assignable to type never' error

**Comment by ardatan** [comment]
Fixed in #1389 
Could you try with these canary versions?
```
Successfully published:
 - graphql-code-generator@0.18.0-alpha.ad875484
 - graphql-codegen-core@0.18.0-alpha.ad875484
 - codegen-templates-scripts@0.18.0-alpha.ad875484
 - graphql-codegen-plugin-handlebars-helpers@0.18.0-alpha.ad875484
 - graphql-codegen-plugin-helpers@0.18.0-alpha.ad875484
 - graphql-codegen-add@0.18.0-alpha.ad875484
 - graphql-codegen-flow-documents@0.18.0-alpha.ad875484
 - graphql-codegen-flow-resolvers@0.18.0-alpha.ad875484
 - graphql-codegen-flow@0.18.0-alpha.ad875484
 - graphql-codegen-fragment-matcher@0.18.0-alpha.ad875484
 - graphql-codegen-introspection@0.18.0-alpha.ad875484
 - graphql-codegen-schema-ast@0.18.0-alpha.ad875484
 - graphql-codegen-time@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-apollo-angular@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-client@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-common@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-graphql-files-modules@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-mongodb@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-react-apollo@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-resolvers@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-server@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-stencil-apollo@0.18.0-alpha.ad875484
 - graphql-codegen-apollo-angular-template@0.18.0-alpha.ad875484
 - graphql-codegen-graphql-files-typescript-modules@0.18.0-alpha.ad875484
 - graphql-codegen-introspection-template@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-mongodb-template@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-react-apollo-template@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-resolvers-template@0.18.0-alpha.ad875484
 - graphql-codegen-typescript-template@0.18.0-alpha.ad875484
lerna success publish finished
```

**Comment by SimenB** [comment]
Thanks! It almost works 🙂 

image

```ts
export type IDirectiveResolvers<Result> = {
  client: ClientDirectiveResolver<Result>;
  skip: SkipDirectiveResolver<Result>;
  include: IncludeDirectiveResolver<Result>;
  deprecated: DeprecatedDirectiveResolver<Result>;
} & { [directiveName: string]: DirectiveResolverFn<any, any, TContext> };
```

**Comment by ardatan** [comment]
Could you try again with `0.18.0-alpha.69f43292`?

**Comment by SimenB** [comment]
The generated code now compiles! I'm having issues passing `TContext` through, though. I have `IResolvers<{ cache: InMemoryCache }>`, but end up with this error:

image

image

**Comment by ardatan** [comment]
@SimenB So, now is it working as expected?

**Comment by SimenB** [comment]
Yes! 🙂

**Comment by ctrlplusb** [comment]
> on TS 3.3.333 these interfaces are now unimplementable, 'not assignable to type never' error.

The `0.19.0-alpha.3973cc8b` canary release resolved this for myself.

**Comment by SimenB** [comment]
Same here, eagerly awaiting the release so we can upgrade

**Comment by Pruxis** [comment]
When I am using 0.19.0-alpha.3973cc8b canary version I am not able to generate my regular types. But the resolvers seem to be correct.

**Pull request merged**